adr/0009-extension-structure.adoc

```diff
@@ -0,0 +1,284 @@
+For example: `org.apache.camel.quarkus.component` would use `org.apache.camel.quarkus.component.runtime`, `org.apache.camel.quarkus.component.dev`, etc.
+====
+
+==== Root package name
+
+To transform an extension name into a package name, use the following pattern:
+
+* `io.quarkus.<extension-name>` for Quarkus core extensions.
+* `io.quarkiverse.<extension-name>` for Quarkiverse extensions.
+* If the extension name contains a hyphen, it is recommended to replace it with an underscore (e.g., `quarkus-foo-bar` becomes `io.quarkus.foo_bar`). 
+
+==== `runtime` module
+* `io.<quarkus|quarkiverse>.<extension-name>.runtime.internal|impl`: Internal implementation. Not part of the public API. `internal` recommended.
+* `io.<quarkus|quarkiverse>.<extension-name>.runtime.graal`: GraalVM substitutions. Not part of the public API.
+* `io.<quarkus|quarkiverse>.<extension-name>.runtime`: Public API when requiring tight-coupling. May include subpackages (excluding `api` and `dev` packages). Example: `io.quarkus.cache`. Note that this is discouraged in favor of the `runtime-api` module.
+
+Non-API packages should be under `internal` like `io.quarkus.<extension-name>.runtime.internal` (`io.quarkus.<extension-name>.runtime.impl` is tolerated).
+Thus, automated tooling can easily use this information to help generate a `module-info` with the correct exports, as well as excluding those packages from JavaDoc generation.
+
+==== `deployment` module
+* `io.<quarkus|quarkiverse>.<extension-name>.deployment`: Internal build logic (processors, build steps, internal build items). Public SPIs must reside in the `deployment-spi` module. May include subpackages (excluding `spi`).
+
+==== `deployment-spi` module
+* `io.<quarkus|quarkiverse>.<extension-name>.deployment.spi`: Public build items and types. Considered public API and subject to compatibility guarantees.
+
+Build items in this module should be documented with their intended use (produced or consumed).
+
+==== `runtime-api` module
+* `io.<quarkus|quarkiverse>.<extension-name>.api`: Public runtime API. Can be used independently of the full extension. Consumers should not expect the full extension to be available at runtime. To make sure the extension is available, the consumer should use the `runtime` module (which would pull in the `runtime-api` module transitively).
+
+==== `runtime-dev` module
+* `io.<quarkus|quarkiverse>.<extension-name>.dev`: Dev-mode-only runtime classes, e.g., for Dev UI contribution. Not included in production builds.
+
+==== `codegen` module
+
+* `io.<quarkus|quarkiverse>.<extension-name>.codegen`: Source code generation logic, if applicable. This module is optional and may not be present in all extensions.
+
+==== `cli` module
+
+* `io.<quarkus|quarkiverse>.<extension-name>.cli`: Quarkus CLI plugin. This module is optional and may not be present in all extensions.
```

Comment:
Does `cli` im/explicitly require `runtime`?
Can CLI plugins also exist in `runtime` (if runtime is small)?